src/modules/musig/main_impl.h

```diff
@@ -0,0 +1,586 @@
+/**********************************************************************
+ * Copyright (c) 2018 Andrew Poelstra                                 *
+ * Distributed under the MIT software license, see the accompanying   *
+ * file COPYING or http://www.opensource.org/licenses/mit-license.php.*
+ **********************************************************************/
+
+#ifndef _SECP256K1_MODULE_MUSIG_MAIN_
+#define _SECP256K1_MODULE_MUSIG_MAIN_
+
+#include "include/secp256k1.h"
+#include "include/secp256k1_musig.h"
+#include "hash.h"
+
+/* Computes ell = SHA256(pk[0], ..., pk[np-1]) */
+static int secp256k1_musig_compute_ell(const secp256k1_context *ctx, unsigned char *ell, const secp256k1_pubkey *pk, size_t np) {
+    secp256k1_sha256 sha;
+    size_t i;
+
+    secp256k1_sha256_initialize(&sha);
+    for (i = 0; i < np; i++) {
+        unsigned char ser[33];
+        size_t serlen = sizeof(ser);
+        if (!secp256k1_ec_pubkey_serialize(ctx, ser, &serlen, &pk[i], SECP256K1_EC_COMPRESSED)) {
+            return 0;
+        }
+        secp256k1_sha256_write(&sha, ser, serlen);
+    }
+    secp256k1_sha256_finalize(&sha, ell);
+    return 1;
+}
+
+/* Compute r = SHA256(ell, idx) */
+static void secp256k1_musig_coefficient(secp256k1_scalar *r, const unsigned char *ell, size_t idx) {
+    secp256k1_sha256 sha;
+    unsigned char buf[32];
+    secp256k1_sha256_initialize(&sha);
+    secp256k1_sha256_write(&sha, ell, 32);
+    while (idx > 0) {
```

Comment:
I don't think it's wrong but it seems somewhat weird not to write leading zeros to the hash. Or is there a good reason, maybe to keep it platform-independent? Just checking if this is intentional.

Reply:
I don't think this has anything to do with platform independence (@apoelstra). I'm happy to fill the remaining bytes with 0's so we're always writing the full size_t.

Reply:
I just noticed that this *is* kind of crucial because we want the MuSig coefficient to be simple and easily implemented by others to come to some kind of interoperable MuSig standard.

Reply:
Sure, we can always write 4 bytes. I did this so I wouldn't have to think about maximum signer size sets.

Reply:
so I'm somewhat siding with the original implementation now. Added a comment `The serialization of idx is least significant byte first and variable-length such that the last byte is non-zero`.

Reply:
... although, if you mess up the serialization you become immediately attackable via key cancellation. So maybe 4 bytes is better. Apparently it's to late for me right now to decide :).

Reply:
Okay, so if Andrew didn't have a strong reason to use this format, I'm fine with both of course.
My slight preference is writing the 4 bytes because it's easier to re-implement (no while loop, no bit fiddling, you typically have some 32-bit type, etc...)

Reply:
using 4 bytes now